### PR TITLE
Add in support for setting delim when parsing JSON through java

### PR DESCRIPTION
**Pull request by revans2**
## Description
This just adds in JNI APIs to allow for changing the delimiter when parsing JSON.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by abellina — APPROVED** [review]

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 1]
```diff
@@ -38,6 +38,7 @@ public final class JSONOptions extends ColumnFilterOptions {
   private final boolean allowLeadingZeros;
   private final boolean allowNonNumericNumbers;
   private final boolean allowUnquotedControlChars;
+  private final byte delim;
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:41]: Let's use the full name `delimiter`.

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 2]
```diff
@@ -52,6 +53,11 @@ private JSONOptions(Builder builder) {
     allowLeadingZeros = builder.allowLeadingZeros;
     allowNonNumericNumbers = builder.allowNonNumericNumbers;
     allowUnquotedControlChars = builder.allowUnquotedControlChars;
+    delim = builder.delim;
+  }
+
+  public byte getDelim() {
+    return delim;
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:60]:
```suggestion
    delim = builder.delim;
  }

  public byte getDelimiter() {
    return delimiter;
```

File: java/src/main/java/ai/rapids/cudf/JSONOptions.java [hunk 3]
```diff
@@ -123,6 +129,16 @@ public static final class Builder  extends ColumnFilterOptions.Builder<JSONOptio
...
 
+    private byte delim = '\n';
+
+    public Builder withDelim(char delimiter) {
+      if (delimiter > Byte.MAX_VALUE) {
+        throw new IllegalArgumentException("Only basic ASCII values are supported " + delimiter);
+      }
+      delim = (byte)delimiter;
+      return this;
+    }
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/JSONOptions.java:140]:
```suggestion
    private byte delimiter = '\n';

    public Builder withDelimiter(char delimiter) {
      if (delimiter > Byte.MAX_VALUE) {
        throw new IllegalArgumentException("Only basic ASCII values are supported " + delimiter);
      }
      this.delimiter = (byte)delimiter;
      return this;
    }
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 4]
```diff
@@ -258,7 +258,8 @@ private static native long readJSON(int[] numChildren, String[] columnNames,
                                         boolean strictValidation,
                                         boolean allowLeadingZeros,
                                         boolean allowNonNumericNumbers,
-                                        boolean allowUnquotedControl) throws CudfException;
+                                        boolean allowUnquotedControl,
+                                        byte delim) throws CudfException;
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:262]:
```suggestion
                                        byte delimiter) throws CudfException;
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 5]
```diff
@@ -272,6 +273,7 @@ private static native long readJSONFromDataSource(int[] numChildren, String[] co
                                       boolean allowLeadingZeros,
                                       boolean allowNonNumericNumbers,
                                       boolean allowUnquotedControl,
+                                      byte delim,
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:276]:
```suggestion
                                      byte delimiter,
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 6]
```diff
@@ -284,6 +286,7 @@ private static native long readAndInferJSONFromDataSource(boolean dayFirst, bool
...
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:289]:
```suggestion
                                      byte delimiter,
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 7]
```diff
@@ -297,7 +300,8 @@ private static native long readAndInferJSON(long address, long length,
                                               boolean strictValidation,
                                               boolean allowLeadingZeros,
                                               boolean allowNonNumericNumbers,
-                                              boolean allowUnquotedControl) throws CudfException;
+                                              boolean allowUnquotedControl,
+                                              byte delim) throws CudfException;
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:304]:
```suggestion
                                              byte delimiter) throws CudfException;
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 8]
```diff
@@ -1321,7 +1325,8 @@ public static Table readJSON(Schema schema, JSONOptions opts, File path) {
                     opts.strictValidation(),
                     opts.leadingZerosAllowed(),
                     opts.nonNumericNumbersAllowed(),
-                    opts.unquotedControlChars()))) {
+                    opts.unquotedControlChars(),
+                    opts.getDelim()))) {
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:1329]:
```suggestion
                    opts.getDelimiter()))) {
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 9]
```diff
@@ -1404,7 +1409,8 @@ public static TableWithMeta readJSON(JSONOptions opts, HostMemoryBuffer buffer,
         opts.strictValidation(),
         opts.leadingZerosAllowed(),
         opts.nonNumericNumbersAllowed(),
-        opts.unquotedControlChars()));
+        opts.unquotedControlChars(),
+        opts.getDelim()));
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:1413]:
```suggestion
        opts.getDelimiter()));
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 10]
```diff
@@ -1426,6 +1432,7 @@ public static TableWithMeta readAndInferJSON(JSONOptions opts, DataSource ds) {
           opts.leadingZerosAllowed(),
           opts.nonNumericNumbersAllowed(),
           opts.unquotedControlChars(),
+          opts.getDelim(),
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:1435]:
```suggestion
          opts.getDelimiter(),
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 11]
```diff
@@ -1479,7 +1486,8 @@ public static Table readJSON(Schema schema, JSONOptions opts, HostMemoryBuffer b
             opts.strictValidation(),
             opts.leadingZerosAllowed(),
             opts.nonNumericNumbersAllowed(),
-            opts.unquotedControlChars()))) {
+            opts.unquotedControlChars(),
+            opts.getDelim()))) {
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:1490]:
```suggestion
            opts.getDelimiter()))) {
```

File: java/src/main/java/ai/rapids/cudf/Table.java [hunk 12]
```diff
@@ -1518,6 +1526,7 @@ public static Table readJSON(Schema schema, JSONOptions opts, DataSource ds, int
         opts.leadingZerosAllowed(),
         opts.nonNumericNumbersAllowed(),
         opts.unquotedControlChars(),
+        opts.getDelim(),
```
Comment (ttnghia) [inline java/src/main/java/ai/rapids/cudf/Table.java:1529]:
```suggestion
        opts.getDelimiter(),
```

File: java/src/test/java/ai/rapids/cudf/TableTest.java [hunk 13]
```diff
@@ -656,6 +656,24 @@ void testJSONValidationUnquotedControl() {
...
 
+  private static final byte[] CR_JSON_TEST_BUFFER = ("{\"a\":\"12\n3\"}\0" +
+      "{\"a\":\"AB\nC\"}\0").getBytes(StandardCharsets.UTF_8);
+
+  @Test
+  void testReadJSONDelim() {
+    Schema schema = Schema.builder().addColumn(DType.STRING, "a").build();
+    JSONOptions opts = JSONOptions.builder()
+        .withLines(true)
+        .withDelim('\0')
```
Comment (ttnghia) [inline java/src/test/java/ai/rapids/cudf/TableTest.java:667]:
```suggestion
        .withDelimiter('\0')
```

**Review by ttnghia — APPROVED** [review]

**Pull request merged**